src/main/java/picard/sam/HitsForInsert.java

```diff
@@ -194,45 +195,100 @@ public void setPrimaryAlignment(final int primaryAlignmentIndex) {
+
+            // Both alignments do not have HI, so push down the ones on the right.
+            // Right is arbitrarily picked to push down.
+            if (isNull(hiTagValueOfFirst) && isNull(hiTagValueOfSecond)) secondOfPair.add(i, null);
         }
 
-        // Now renumber any correlated alignments, and remove hit index if no correlated read.
-        int hi = 0;
+        renumberHitIndex();
+    }
+
+    /**
+     * Renumber any correlated alignments, and remove hit index if no correlated read.
+     */
+    private void renumberHitIndex() {
         for (int i = 0; i < numHits(); ++i) {
             final SAMRecord first = getFirstOfPair(i);
             final SAMRecord second = getSecondOfPair(i);
-            if (first != null && second != null) {
+            if (nonNull(first) && nonNull(second)) {
                 first.setAttribute(SAMTag.HI.name(), i);
                 second.setAttribute(SAMTag.HI.name(), i);
-                ++hi;
-            } else if (first != null) {
-                first.setAttribute(SAMTag.HI.name(), null);
-            } else {
-                second.setAttribute(SAMTag.HI.name(), null);
+            }
+            if (nonNull(first) && isNull(second)) first.setAttribute(SAMTag.HI.name(), null);
+            if (isNull(first) && nonNull(second)) second.setAttribute(SAMTag.HI.name(), null);
+        }
+    }
+
+    /**
+     * This method lines up alignments in firstOfPairOrFragment and secondOfPair lists so that the first and the second
+     * records of each pair have the same index.
+     */
+    void coordinateByMate() {
+        final List<SAMRecord> newFirstOfPairOrFragment = new ArrayList<>();
+        final List<SAMRecord> newSecondOfPair = new ArrayList<>();
+
+        for (int i = 0; i < firstOfPairOrFragment.size(); i++) {
+            final SAMRecord first = firstOfPairOrFragment.get(i);
+            newFirstOfPairOrFragment.add(i, first);
+            newSecondOfPair.add(null);
+
+            for (int k = 0; k < secondOfPair.size(); k++) {
+                final SAMRecord second = secondOfPair.get(k);
+                if (arePair(first, second)) {
+                    newSecondOfPair.set(i, second);
+                    secondOfPair.set(k, null);
+                    break;
+                }
+            }
+        }
+
+        for (int i = 0; i < secondOfPair.size(); i++) {
+            if (nonNull(secondOfPair.get(i))) {
+                newFirstOfPairOrFragment.add(i, null);
+                newSecondOfPair.add(i, secondOfPair.get(i));
             }
         }
+
+        firstOfPairOrFragment.clear();
+        firstOfPairOrFragment.addAll(newFirstOfPairOrFragment);
+        secondOfPair.clear();
+        secondOfPair.addAll(newSecondOfPair);
+
+        renumberHitIndex();
+    }
+
+    /**
+     * @return true if records belong to the same pairwise alignment
+     */
+    private static boolean arePair(SAMRecord first, SAMRecord second) {
```

Comment:
This looks like it should be a non-static method on `SAMRecord`, as it only deals with fields in that object and takes that object as its argument.

Reply:
`SAMRecord` is part of `htsjdk` and this was the only constraint I haven't done it. Could I just make this function a `SamRecord#arePair` method and make associated pull-request to [`htsjdk`](https://github.com/samtools/htsjdk)?

Reply:
Could you please review [associated request](https://github.com/samtools/htsjdk/pull/994) to `htsjdk`?